docs/agents/span-min-duration.md

```diff
@@ -0,0 +1,126 @@
+# Discarding fast spans
+
+Currently,
+the experience of our APM solution is not ideal when monitoring an application that ends up generating thousands of spans per request.
+
+We currently have a default limit of 500 spans in the agents and 1000 spans in the UI to prevent overwhelming the system with too many spans.
+However,
+that just cuts off at a certain point without considering which spans are important and which are not.
+
+## Configuration option
+
+Key: `span_min_duration`
+
+Default value: 0ms (don't discard any spans)
+
+Description:
+Sets the minimum duration of spans.
+Spans that execute faster than this threshold are attempted to be discarded.
+
+The attempt fails if they lead up to a span that can't be discarded.
+Spans that propagate the trace context to downstream services,
+such as outgoing HTTP requests,
+can't be discarded.
+Additionally, spans that lead to an error or that may be a parent of an async operation can't be discarded.
+
+However, external calls that don't propagate context,
+such as calls to a database, can be discarded using this threshold.
+
+## Limitations
+
+The limitations are based on the premise that the `parent_id` of each span and transaction that's stored in Elasticsearch
+should point to another valid transaction or span that's present in the Elasticsearch index.
+
+A span that refers to a missing span via is `parent_id` is also known as an "orphaned span".
+
+### Spans that propagate context to downstream services can't be discarded
+
+We only know whether to discard after the call has ended.
+At that point,
+the trace has already continued on the downstream service.
+Discarding the span for the external request would orphan the transaction of the downstream call.
+
+Propagating the trace context to downstream services is also known as out-of-process context propagation.
+
+### Intermediate spans
+
+Discarding a single span that has both a parent and a child span is not possible as it would lead to orphaned child spans.
+
+However,
+a whole subtree of spans can get discarded if all spans within that tree are requesting to be discarded.
+This means that if a leaf of the tree can't be discarded because it propagates context downstream,
+all spans leading up to it can't be discarded.
+If the leaf is a non-context propagating span,
+such as a manually created span or a SQL call,
+the subtree can be discarded.
+
+### Spans that lead up to async spans can't be discarded
```

Comment:
This limitation might make the proposal not very useful for Node.js or Go/goroutines.

Reply:
See also @axw's comment from https://github.com/elastic/apm/issues/259#issuecomment-617579842

> I don't have a specific idea of how it would work yet, I just know that detecting transfer of context between threads is not a viable approach in Go. This bit is pretty language specific I think. For Go I would probably make it so that spans are assumed to be synchronous by default, and callers of the API could explicitly mark them as async.